crates/uv-publish/src/lib.rs

```diff
@@ -369,26 +383,123 @@ pub async fn upload(
             )
         })?;
 
-        return handle_response(registry, response)
+        return match handle_response(registry, response).await {
+            Ok(()) => {
+                // Upload successful; for PyPI this can also mean a hash match in a raced upload
+                // (but it doesn't tell us), for other registries it should mean a fresh upload.
+                Ok(true)
+            }
+            Err(err) => {
+                if matches!(
+                    err,
+                    PublishSendError::Status(..) | PublishSendError::StatusNoBody(..)
+                ) && check_url(index_client, index_capabilities, file, filename).await?
+                {
+                    // There was a raced upload of the same file, so even though our upload failed,
+                    // the right file now exists in the registry.
+                    return Ok(false);
+                }
+                Err(PublishError::PublishSend(
+                    file.to_path_buf(),
+                    registry.clone(),
+                    err,
+                ))
+            }
+        };
+    }
+}
+
+/// Check whether we should skip the upload of a file because it already exists on the index.
+pub async fn check_url(
+    index_client: &mut Option<(IndexUrl, RegistryClient)>,
+    index_capabilities: &IndexCapabilities,
+    file: &Path,
+    filename: &DistFilename,
+) -> Result<bool, PublishError> {
+    let Some((index_url, registry_client)) = index_client.as_mut() else {
+        return Ok(false);
+    };
```

Comment:
I think I'd prefer that this method _not_ take an `Option`, and we just don't call it if `index_client` is `None`.